src/core/http1/client.rs

```diff
@@ -0,0 +1,604 @@
+        self,
+        method: &str,
+        uri: &str,
+        headers: &[http1::Header<'_>],
+        body_size: http1::BodySize,
+        websocket: bool,
+        initial_body: &[u8],
+        end: bool,
+    ) -> Result<RequestHeader<'a, R, W>, Error> {
+        let req = http1::ClientRequest::new();
+
+        let req_body = match req.send_header(self.buf1, method, uri, headers, body_size, websocket)
+        {
+            Ok(ret) => ret,
+            Err(_) => return Err(Error::BufferExceeded),
+        };
+
+        if self.buf2.write_all(initial_body).is_err() {
+            return Err(Error::BufferExceeded);
+        }
+
+        Ok(RequestHeader {
+            r: self.r,
+            w: self.w,
+            buf1: self.buf1,
+            buf2: self.buf2,
+            req_body,
+            end,
+        })
+    }
+}
+
+pub struct RequestHeader<'a, R: AsyncRead, W: AsyncWrite> {
+    r: ReadHalf<'a, R>,
+    w: WriteHalf<'a, W>,
+    buf1: &'a mut VecRingBuffer,
+    buf2: &'a mut VecRingBuffer,
+    req_body: http1::ClientRequestBody,
+    end: bool,
+}
+
+impl<'a, R: AsyncRead, W: AsyncWrite> RequestHeader<'a, R, W> {
+    pub async fn send(mut self) -> Result<RequestBody<'a, R, W>, Error> {
+        while self.buf1.len() > 0 {
+            let size = self.w.write(Buffer::read_buf(self.buf1)).await?;
+            self.buf1.read_commit(size);
+        }
+
+        let block_size = self.buf2.capacity();
+
+        Ok(RequestBody {
+            inner: RefCell::new(Some(RequestBodyInner {
+                r: RefCell::new(RequestBodyRead {
+                    stream: self.r,
+                    buf: self.buf1,
+                }),
+                w: RefCell::new(RequestBodyWrite {
+                    stream: self.w,
+                    buf: self.buf2,
+                    req_body: Some(self.req_body),
+                    end: self.end,
+                    block_size,
+                }),
+            })),
+        })
+    }
+}
+
+struct RequestBodyRead<'a, R: AsyncRead> {
+    stream: ReadHalf<'a, R>,
+    buf: &'a mut VecRingBuffer,
+}
+
+struct RequestBodyWrite<'a, W: AsyncWrite> {
+    stream: WriteHalf<'a, W>,
+    buf: &'a mut VecRingBuffer,
+    req_body: Option<http1::ClientRequestBody>,
+    end: bool,
+    block_size: usize,
+}
+
+struct RequestBodyInner<'a, R: AsyncRead, W: AsyncWrite> {
+    r: RefCell<RequestBodyRead<'a, R>>,
+    w: RefCell<RequestBodyWrite<'a, W>>,
+}
+
+pub struct RequestBody<'a, R: AsyncRead, W: AsyncWrite> {
+    inner: RefCell<Option<RequestBodyInner<'a, R, W>>>,
+}
+
+impl<'a, R: AsyncRead, W: AsyncWrite> RequestBody<'a, R, W> {
+    pub fn prepare(&self, src: &[u8], end: bool) -> Result<usize, Error> {
+        if let Some(inner) = &*self.inner.borrow() {
+            let w = &mut *inner.w.borrow_mut();
+
+            // call not allowed if the end has already been indicated
+            if w.end {
+                return Err(Error::Io(io::Error::from(io::ErrorKind::InvalidInput)));
```

Comment:
this errkind doesn't seem to match the case. maybe sth like `NoFurtherWritesAllowed` or 
return Err(Error::Io(io::Error::new(io::ErrorKind::Other, "No further writes allowed, end flag set")));